src/System.Management.Automation/engine/ExecutionContext.cs

```diff
@@ -1269,55 +1269,116 @@ internal PSTransactionManager TransactionManager
 
         internal PSTransactionManager transactionManager;
 
-        internal Assembly AddAssembly(string name, string filename, out Exception error)
-        {
-            Assembly loadedAssembly = LoadAssembly(name, filename, out error);
-
-            if (loadedAssembly == null)
-                return null;
-
-            if (AssemblyCache.ContainsKey(loadedAssembly.FullName))
+        /// <summary>
+        /// This method is used for assembly loading requests stemmed from 'InitialSessionState' binding and module loading.
+        /// </summary>
+        /// <param name="source">Source of the assembly loading request, usually is a module name.</param>
+        /// <param name="assemblyName">Name of the assembly to be loaded.</param>
+        /// <param name="fileName">Path of the assembly to be loaded.</param>
+        /// <param name="error">Exception that is caught when the loading fails.</param>
+        internal Assembly AddAssembly(string source, string assemblyName, string fileName, out Exception error)
+        {
+            // Search the cache by the path, and return the assembly if we find it.
+            // It's common to have two loading requests for the same assembly when loading a module -- the first time for
+            // resolving a binary module path, and the second time for actually processing that module.
+            //
+            // That's not a problem when all the module assemblies are loaded into the default ALC. But in a scenario where
+            // a module tries to hide its nested/root binary modules in a custom ALC, that will become a problem.
+            //
+            // In this scenario, the module will usually setup a handler to load the specific assemblies to the custom ALC,
+            // and that will be how the first loading request gets served. However, after the module path is resolved with
+            // the first loading, the path will be used for the second loading upon real module processing. Since we prefer
+            // loading by path to loading by name, we will end up loading the same assembly in the default ALC if we do not
```

Comment:
I don't understand this sentence.  If the assembly `filePath` uniquely defines the assembly for both loading and cache look-up, why does it matter to check the cache first?

It looks like this is because the cache now includes `source` along with `filePath` to uniquely identify the assembly.  These comments should make clear that assembly `load` and `cache lookup` differ in this way.  Also these comments should make it clear that the `second` assembly load bypasses the `resolver` handler, I guess because it has its own cache look up that returns from the default ALC (incorrectly).

Reply:
The purpose of checking cache first is to prevent the same assembly file from being loaded into the default ALC when it's supposed to be loaded only by the custom ALC from the module.

`ExecutionContext.LoadAssembly` uses `Assembly.LoadFrom` for loading from assembly path. That API always loads an assembly to the default ALC. For the scenario this PR tries to address,
- the first loading request (for resolving the assembly path) will be handled by the module's custom ALC -- the assembly will be loaded by the custom ALC and put in the cache. The assembly path will be resolved to be `assembly.Location`.
- the second loading request comes with the resolved assembly path. If we don't check the cache first, we will call into `LoadAssembly -> Assembly.LoadFrom(filePath)` and load that same assembly into the default ALC -- the scenario will be broken. This is why we need to check the cache first.

As for `source`, it's mainly just for removing the cached assemblies when a module is unloaded. The module info doesn't keep track of all resolved module assembly paths, and it would be a non-trivial change to track them all. Having the source as the prefix to the key would make it easy to clean up all assemblies loaded by a module from the cache.

In practice, I don't think it's possible to have the same assembly path with different source prefix in the cache. It could happen in theory, where 2 different modules using the same GAC'ed assembly as a module assembly, but even that should be fine.

Reply:
Ok, this makes sense.  I was wondering why the filePath would not be unique even for assemblies in a different ALC (different location).  Maybe emphasize in the comments that this is to prevent loading an isolated assembly into the default ALC when using the full filePath.

So this scheme really depends on module assembly loading always following the same pattern:  1.  loadByName (resolver loads into correct ALC)  2.  loadByCache or loadByFullFilePath (always loads into default ALC).  Maybe include a simple example of this pattern in the comments.

Do we know this sequence is always followed?  Also we should comment where in module loading this sequence is performed and include a comment explaining how this should never change as it would break custom ALC modules.

Reply:
The comments were updated with an example.